File: Arrays & HashMap/Group Anagrams/Group Anagrams.kt
class Solution {
    fun groupAnagrams(strs: Array<String>): List<List<String>> {
        val map = mutableMapOf<MutableList<Int>, MutableList<String>>()

        strs.forEach { str -> 
            val count = MutableList(26) {0}
            str.forEach { c -> 
                count[c.code - 'a'.code] += 1
            }
            val tempList = map.getOrDefault(count, mutableListOf())
            tempList.add(str)
            map[count] = tempList
        }
        return map.values.toList()
    }
}

//Time Complexity: O(n)
//Space Complexity: O(n)

File: Arrays & HashMap/Group Anagrams/Group Anagrams.py
class Solution:
    def groupAnagrams(self, words: List[str]) -> List[List[str]]:
        map = defaultdict(list)

        for word in words:
            chars = [0] * 26
            for char in word:
                chars[ord(char) - ord('a')] += 1
            
            
            map[tuple(chars)].append(word)

        return list(map.values())

'''
Time Complexity: O(n)
Space Complexity: O(n)
'''